packages/fauna/src/index.ts

```diff
@@ -0,0 +1,216 @@
+  to(object: Record<string, any>) {
+    const newObject: Record<string, unknown> = {}
+    for (const key in object) {
+      const value = object[key]
+      if (value instanceof Date) {
+        newObject[key] = Time(value.toISOString())
+      } else {
+        newObject[key] = value
+      }
+    }
+    return newObject
+  },
+  /** Takes a Fauna object and returns a plain old JavaScript object */
+  from<T = Record<string, unknown>>(object: Record<string, any>): T {
+    const newObject: Record<string, unknown> = {}
+    for (const key in object) {
+      const value = object[key]
+      if (value?.value) {
+        newObject[key] = new Date(value.value)
+      } else {
+        newObject[key] = value
+      }
+    }
+    return newObject as T
+  },
+}
+
+/**
+ * FaunaDB throws an error when something is not found in the db,
+ * `next-auth` expects `null` to be returned
+ */
+export function query(client: FaunaClient, format: (...args: any) => any) {
+  return async function <T>(expr: ExprArg): Promise<T | null> {
+    try {
+      const result = await client.query<QueryResult<T> | null>(expr)
+      if (!result) return null
+      return format({ ...result.data, id: result.ref.id })
+    } catch (error) {
+      if (error.name === "NotFound") return null
+      if (error.description?.includes("Number or numeric String expected"))
+        return null
+
+      if (process.env.NODE_ENV === "test") console.error(error)
+
+      throw error
+    }
+  }
+}
+
+export function FaunaAdapter(f: FaunaClient): Adapter {
+  const { Users, Accounts, Sessions, VerificationTokens } = collections
+  const { to, from } = format
+  const q = query(f, from)
+  return {
+    createUser: async (data) => (await q(Create(Users, { data: to(data) })))!,
+    getUser: async (id) => await q(Get(Ref(Users, id))),
+    getUserByEmail: async (email) =>
+      await q(Get(Match(indexes.UserByEmail, email))),
+    async getUserByAccount({ provider, providerAccountId }) {
+      const key = [provider, providerAccountId]
+      const ref = Match(indexes.UserByAccount, key)
+      const user = await q<AdapterUser>(
+        Let(
+          { ref },
+          If(
+            Exists(Var("ref")),
+            Get(Ref(Users, Select(["data", "userId"], Get(Var("ref"))))),
+            null
+          )
+        )
+      )
+      return user
+    },
+    updateUser: async (data) =>
+      (await q(Update(Ref(Users, data.id), { data: to(data) })))!,
+    async deleteUser(userId) {
+      await f.query(
+        Let(
+          {
+            deleteSessions: Map(
+              Paginate(Match(indexes.SessionsByUser, userId)),
+              Lambda("ref", Delete(Var("ref")))
+            ),
+            deleteAccounts: Map(
+              Paginate(Match(indexes.AccountsByUser, userId)),
+              Lambda("ref", Delete(Var("ref")))
+            ),
+            user: Delete(Ref(Users, userId)),
+          },
+          true
+        )
```

Comment:
I think this is fine, but since we don't need to care about the return value here, we can maybe use `Do` and `ForEach` instead of `Let` and `Map`. This is just a suggestion. What you already have is fine too.

Reply:
Could you leave a code suggestion?

Reply:
```suggestion
        Do(
            Foreach(
              Paginate(Match(indexes.SessionsByUser, userId)),
              Lambda("ref", Delete(Var("ref")))
            ),
            Foreach(
              Paginate(Match(indexes.AccountsByUser, userId)),
              Lambda("ref", Delete(Var("ref")))
            ),
            Delete(Ref(Users, userId))
        )
```